edb/pgsql/compiler/relgen.py

```diff
@@ -3483,6 +3483,45 @@ def process_set_as_exists_expr(
     return new_stmt_set_rvar(ir_set, stmt, ctx=ctx)
 
 
+@_special_case('std::json_object_pack')
+def process_set_as_json_object_pack(
+        ir_set: irast.Set, stmt: pgast.SelectStmt, *,
+        ctx: context.CompilerContextLevel) -> SetRVars:
+    expr = ir_set.expr
+    assert isinstance(expr, irast.FunctionCall)
+
+    with ctx.subrel() as subctx:
+        ir_arg = expr.args[0].expr
+
+        # compile IR to pgast that may be a ref
+        arg_ref = dispatch.compile(ir_arg, ctx=subctx)
+
+        # ensure that pgast ref is actually a value
+        arg_val = output.output_as_value(arg_ref, env=subctx.env)
```

Comment:
A couple things:
1. Apparently `compile` always calls `output_as_value` on its result, so we shouldn't need to call `output_as_value`. (There are a bunch of existing call sites that I think do anyway.)
2. Let's try handling the special case (though actually pretty common; I checked, and an `array_unpack` of an array of tuples will do this) case of having a `TupleVar`. To do this, you're going to want to call `visit` instead of `compile` (so it skips computing the value), grab the path var out yourself with `get_path_value_var`, and then check on the result to see if it is a `TupleVar`